roofit/roofitcore/src/RooNLLVar.cxx

```diff
@@ -465,28 +466,37 @@ RooNLLVar::ComputeResult RooNLLVar::computeBatchedFunc(const RooAbsPdf *pdfClone
 
 RooNLLVar::ComputeResult RooNLLVar::computeScalar(std::size_t stepSize, std::size_t firstEvent, std::size_t lastEvent) const {
   auto pdfClone = static_cast<const RooAbsPdf*>(_funcClone);
-  return computeScalarFunc(pdfClone, _dataClone, _normSet, _weightSq, stepSize, firstEvent, lastEvent);
+  return computeScalarFunc(pdfClone, _dataClone, _normSet, _weightSq, stepSize, firstEvent, lastEvent, _templateRatioOffset);
 }
 
 // static function, also used from TestStatistics::RooUnbinnedL
 RooNLLVar::ComputeResult RooNLLVar::computeScalarFunc(const RooAbsPdf *pdfClone, RooAbsData *dataClone,
                                                       RooArgSet *normSet, bool weightSq, std::size_t stepSize,
-                                                      std::size_t firstEvent, std::size_t lastEvent)
+                                                      std::size_t firstEvent, std::size_t lastEvent, bool templateRatioOffset)
 {
   ROOT::Math::KahanSum<double> kahanWeight;
   ROOT::Math::KahanSum<double> kahanProb;
   RooNaNPacker packedNaN(0.f);
+  const double logSumW = std::log(weightSq ? dataClone->sumEntriesW2() : dataClone->sumEntries());
+
+  auto* dataHist = templateRatioOffset ? dynamic_cast<RooDataHist*>(dataClone) : nullptr;
 
   for (auto i=firstEvent; i<lastEvent; i+=stepSize) {
     dataClone->get(i) ;
 
-    double eventWeight = dataClone->weight(); //FIXME
-    if (0. == eventWeight * eventWeight) continue ;
-    if (weightSq) eventWeight = dataClone->weightSquared() ;
+    double weight = dataClone->weight(); //FIXME
+    if (0. == weight * weight) continue ;
+    if (weightSq) weight = dataClone->weightSquared() ;
 
-    const double term = -eventWeight * pdfClone->getLogVal(normSet);
+    const double logProba = pdfClone->getLogVal(normSet);
+    double term = 0.0;
+    if(templateRatioOffset && dataHist) {
+      term = -weight * (logProba - (std::log(weight) - std::log(dataHist->binVolume(i)) - logSumW));
```

Comment:
Why are you subtracting logSumW here ? I would do outside the loop. Also it is `log ( binVolume * pdf / weight)  that is close to zero (at least in a perfect fit). 

Also in case of weighted fit, when weightSq=true, I think it should be: 
```
term = -weightSquared * (logProba - (std::log(weight) - std::log(dataHist->binVolume(i)))
```
the minimum will be the same, but numerically the above expression will be better

Reply:
Thanks you are right about the weightSq! As for subtracting logSumW: I discussed together with Lorenzo on the whiteboard that we should indeed subtract it in the loop to get terms close to zero.